panther_battery/test/test_roboteq_republisher_node.cpp

```diff
@@ -0,0 +1,344 @@
+  EXPECT_TRUE(std::isnan(battery_state_->temperature));
+  EXPECT_TRUE(CheckNaNVector(battery_state_->cell_voltage));
+  EXPECT_TRUE(CheckNaNVector(battery_state_->cell_temperature));
+  EXPECT_EQ(BatteryStateMsg::POWER_SUPPLY_TECHNOLOGY_LIPO, battery_state_->power_supply_technology);
+
+  // variable values
+  EXPECT_FLOAT_EQ(expected_voltage, battery_state_->voltage);
+  EXPECT_FLOAT_EQ(expected_current, battery_state_->current);
+  EXPECT_FLOAT_EQ(expected_percentage, battery_state_->percentage);
+  EXPECT_FLOAT_EQ(expected_percentage * 20.0, battery_state_->charge);
+
+  if (power_supply_status == BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING) {
+    EXPECT_TRUE(battery_state_->present);
+  }
+  EXPECT_EQ(power_supply_status, battery_state_->power_supply_status);
+  EXPECT_EQ(power_supply_health, battery_state_->power_supply_health);
+}
+
+bool TestRoboteqRepublisherNode::CheckNaNVector(const std::vector<float> & vector)
+{
+  return std::all_of(
+    vector.begin(), vector.end(), [](const float value) { return std::isnan(value); });
+}
+
+TEST_F(TestRoboteqRepublisherNode, BatteryMsgDefaultValues)
+{
+  ASSERT_TRUE(panther_utils::test_utils::WaitForMsg(
+    roboteq_republisher_node_, battery_state_, std::chrono::milliseconds(1000)));
+
+  CheckBatteryStateMsg(
+    0.0, 0.0, 0.0, BatteryStateMsg::POWER_SUPPLY_STATUS_UNKNOWN,
+    BatteryStateMsg::POWER_SUPPLY_HEALTH_UNKNOWN);
+}
+
+TEST_F(TestRoboteqRepublisherNode, BatteryMsgValues)
+{
+  DriverStateMsg driver_state_msg;
+  driver_state_msg.front.voltage = 37.0;
+  driver_state_msg.rear.voltage = 39.0;
+  driver_state_msg.front.current = 0.1;
+  driver_state_msg.rear.current = 0.1;
+
+  driver_state_pub_->publish(driver_state_msg);
+  ASSERT_TRUE(panther_utils::test_utils::WaitForMsg(
+    roboteq_republisher_node_, battery_state_, std::chrono::milliseconds(1000)));
+
+  auto expected_voltage = (driver_state_msg.front.voltage + driver_state_msg.rear.voltage) / 2.0;
+  auto expected_current = driver_state_msg.front.current + driver_state_msg.rear.current;
+  auto expected_percentage = (expected_voltage - 32.0) / (41.4 - 32.0);
+  CheckBatteryStateMsg(
+    expected_voltage, expected_current, expected_percentage,
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD);
+
+  driver_state_msg.front.voltage = 33.0;
+  driver_state_msg.rear.voltage = 35.0;
+  driver_state_msg.front.current = -0.2;
+  driver_state_msg.rear.current = -0.2;
+
+  driver_state_pub_->publish(driver_state_msg);
+  ASSERT_TRUE(panther_utils::test_utils::WaitForMsg(
+    roboteq_republisher_node_, battery_state_, std::chrono::milliseconds(1000)));
+
+  expected_voltage = 36.0;
+  expected_current = -0.1;
+  expected_percentage = 0.425531915;
+  CheckBatteryStateMsg(
+    expected_voltage, expected_current, expected_percentage,
+    BatteryStateMsg::POWER_SUPPLY_STATUS_DISCHARGING, BatteryStateMsg::POWER_SUPPLY_HEALTH_GOOD);
```

Comment:
I am not sure if this shouldn't be a separate test case for "manual" battery percentage. Also the negative current might be a separate test since it is independent to percentage, hence it should be detected by separate test case

Reply:
As discussed, we leave it in 1 test case

Reply:
Ok